packages/video_player/video_player_avfoundation/darwin/video_player_avfoundation/Sources/video_player_avfoundation/FVPVideoPlayerPlugin.m

```diff
@@ -62,32 +81,44 @@ - (instancetype)initWithAVFactory:(id<FVPAVFactory>)avFactory
   _registrar = registrar;
   _displayLinkFactory = displayLinkFactory ?: [[FVPDefaultDisplayLinkFactory alloc] init];
   _avFactory = avFactory ?: [[FVPDefaultAVFactory alloc] init];
-  _playersByTextureId = [NSMutableDictionary dictionaryWithCapacity:1];
+  _playersById = [NSMutableDictionary dictionaryWithCapacity:1];
+  // Initialized to a high number to avoid collisions with texture IDs (which are generated
+  // separately).
+  _nextNonTexturePlayerId = INT_MAX;
   return self;
 }
 
 - (void)detachFromEngineForRegistrar:(NSObject<FlutterPluginRegistrar> *)registrar {
-  [self.playersByTextureId.allValues makeObjectsPerformSelector:@selector(disposeSansEventChannel)];
-  [self.playersByTextureId removeAllObjects];
+  [self.playersById.allValues makeObjectsPerformSelector:@selector(disposeSansEventChannel)];
+  [self.playersById removeAllObjects];
   SetUpFVPAVFoundationVideoPlayerApi(registrar.messenger, nil);
 }
 
-- (int64_t)onPlayerSetup:(FVPVideoPlayer *)player frameUpdater:(FVPFrameUpdater *)frameUpdater {
-  int64_t textureId = [self.registry registerTexture:player];
-  frameUpdater.textureId = textureId;
+- (int64_t)onPlayerSetup:(FVPVideoPlayer *)player {
+  BOOL textureBased = [player isKindOfClass:[FVPTextureBasedVideoPlayer class]];
+  int64_t playerId;
+  if (textureBased) {
```

Comment:
Optional: an alternate way to structure this would be:

```objc
FVPTextureBasedVideoPlayer* textureBasedPlayer = [player isKindOfClass:[FVPTextureBasedVideoPlayer class]] ?
    (FVPTextureBasedVideoPlayer *)player : nil;

if (textureBasedPlayer) {
  playerId = [self.registry registerTexture:textureBasedPlayer];
  ...
}
...
[textureBasedPlayer expectFrame];
```

Then the only casting is happening in the same statement with the check that makes the cast safe, so there's less room for error in future refactoring.